tools/gfx/cuda/render-cuda.cpp

```diff
@@ -0,0 +1,1506 @@
+                cuTexObjectCreate(&tex->m_cudaTexObj, &resDesc, &texDesc, nullptr));
+        }
+
+        *outResource = tex.detach();
+        return SLANG_OK;
+    }
+
+    virtual Result createBufferResource(
+        Resource::Usage initialUsage,
+        const BufferResource::Desc& desc,
+        const void* initData,
+        BufferResource** outResource) override
+    {
+        RefPtr<MemoryCUDAResource> resource = new MemoryCUDAResource(desc);
+        SLANG_CUDA_RETURN_ON_FAIL(cudaMallocManaged(&resource->m_cudaMemory, desc.sizeInBytes));
+        if (initData)
+        {
+            SLANG_CUDA_RETURN_ON_FAIL(cudaMemcpy(resource->m_cudaMemory, initData, desc.sizeInBytes, cudaMemcpyHostToDevice));
+        }
+        *outResource = resource.detach();
+        return SLANG_OK;
+    }
+
+    virtual Result createTextureView(
+        TextureResource* texture, ResourceView::Desc const& desc, ResourceView** outView) override
+    {
+        RefPtr<CUDAResourceView> view = new CUDAResourceView();
+        view->desc = desc;
+        view->textureResource = dynamic_cast<TextureCUDAResource*>(texture);
+        *outView = view.detach();
+        return SLANG_OK;
+    }
+
+    virtual Result createBufferView(
+        BufferResource* buffer, ResourceView::Desc const& desc, ResourceView** outView) override
+    {
+        RefPtr<CUDAResourceView> view = new CUDAResourceView();
+        view->desc = desc;
+        view->memoryResource = dynamic_cast<MemoryCUDAResource*>(buffer);
+        *outView = view.detach();
+        return SLANG_OK;
+    }
+
+    virtual Result createShaderObjectLayout(
+        slang::TypeLayoutReflection* typeLayout, ShaderObjectLayout** outLayout) override
+    {
+        RefPtr<CUDAShaderObjectLayout> cudaLayout;
+        cudaLayout = new CUDAShaderObjectLayout(typeLayout);
+        *outLayout = cudaLayout.detach();
+        return SLANG_OK;
+    }
+
+    virtual Result createRootShaderObjectLayout(
+        slang::ProgramLayout* layout, ShaderObjectLayout** outLayout) override
+    {
+        RefPtr<CUDAProgramLayout> cudaLayout;
+        cudaLayout = new CUDAProgramLayout(layout);
+        cudaLayout->programLayout = layout;
+        *outLayout = cudaLayout.detach();
+        return SLANG_OK;
+    }
+
+    virtual Result createShaderObject(ShaderObjectLayout* layout, ShaderObject** outObject) override
+    {
+        RefPtr<CUDAShaderObject> result = new CUDAShaderObject();
+        SLANG_RETURN_ON_FAIL(result->init(this, dynamic_cast<CUDAShaderObjectLayout*>(layout)));
+        *outObject = result.detach();
+        return SLANG_OK;
+    }
+
+    virtual Result createRootShaderObject(ShaderObjectLayout* layout, ShaderObject** outObject) override
+    {
+        RefPtr<CUDARootShaderObject> result = new CUDARootShaderObject();
+        SLANG_RETURN_ON_FAIL(result->init(this, dynamic_cast<CUDAShaderObjectLayout*>(layout)));
+        *outObject = result.detach();
+        return SLANG_OK;
+    }
```

Comment:
I may have missed it in the previous change that refactored the `ShaderObject` API to be part of `Renderer`, but it seems like we may want to think about tweaking the "root shader object" idea in a few ways:

* It seems like a `Program` that is being loaded/created from Slang could eventually store the `slang::ProgramLayout` directly, at which point any API that currently takes a `slang::ProgramLayout` could just take a `gfx::Program`.

* It seems like it is redundant to have a `createRootShaderObjectLayout` step, because if a `gfx::Program` stores the `slang::ProgramLayout`, then it should also be able to create and store the `gfx`-level "root layout."

* At that point `createRootShaderObject` can just take the `gfx::Program` instead of a `ShaderObjectLayout`.

* At that point, it makes sense that `createRootShaderObject` returns a distinguished sub-type of `ShaderObject` to represent the fact that it is representing a `ProgramObject` or `RootShaderObject` or whatever we want to call it.

* If a `ComputePipelineState` only stores a `Program` (since the layout is now determined by the program), there is no reason to have `ComputePipelineState`.

* If a `RootShaderObject`/`ProgramObject` is always allocated from a `Program` then, binding a root shader object already determined the program to use, so that there need not be distinct calls to set the program and the root shader object.

* If a `RootShaderObject`/`ProgramObject` fully determines the program and the argument data to use for  dispatch, then it seems reasonable to have an overload of `dispatchCompute` that directly takes the root shader object to use for the dispatch, and is thus completely stateless.

One additional longer-term note is that somebody might look at the current `gfx` abstraction and argue that it is inefficient because you end up having to bind all the "root" parameters into an explicit root `ShaderObject`, and *then* bind it with `bindRootShaderObject`, which has the effect of copying state from the `ShaderObject` over to the "real" pipeline/context state (e.g., the copying of data from a `ShaderObject` to the `SLANG_globalParams`, which is equivalent to copying bindings into root parameters).

What I would advocate as a potential long-term solution there is to have an interface where intead of binding a specific root `ShaderObject` (that would have already been filled in), the application binds a specific `Program` and the operation returns a "transient root shader object":

```
RootShaderObject* bindTransientRootShaderObject(Program* program);
```

The user-facing constraint for a "transient root shader object" would be that as soon as another root program/object gets bound, the transient object is no longer valid to use, and is subject to cleanup on the GPU timeline.

Because of that constraint, the implementation of a transient root shader object can know/assume that it is currently bound to a particular context/command-buffer, and that there is no need to persist its state after it get unbound. A dedicated root shader object implementation could thus map operations like `setResource` directly to low-level graphics API calls that bind things to root parameters.

Implementing a dedicated "transient root shader object" abstraction would allow this API layer to potentially get some of the benefits of the API approach advocated for in the "shader components" paper.